src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.pi.lex
+import data.finsupp.order
+import data.finsupp.witnesses
+
+/-!
+# Lexicographic order on finitely supported functions
+
+This file defines the lexicographic order on `finsupp`.
+-/
+
+variables {α N : Type*}
+namespace finsupp
+
+section N_has_zero
+variables [has_zero N]
+
+/-- The lexicographic relation on `α →₀ N`, where `α` is ordered by `r`,
+  and `N` is ordered by `s`. -/
+protected def lex (r : α → α → Prop) (s : N → N → Prop) (x y : α →₀ N) : Prop :=
+pi.lex r (λ _, s) x y
+
+instance [has_lt α] [has_lt N] : has_lt (lex (α →₀ N)) := ⟨finsupp.lex (<) (<)⟩
+
+instance lex.is_strict_order [linear_order α] [partial_order N] :
+  is_strict_order (lex (α →₀ N)) (<) :=
+let i : is_strict_order (lex (α → N)) (<) := pi.lex.is_strict_order in
+{ irrefl := to_lex.surjective.forall.2 $ λ a, @irrefl _ _ i.to_is_irrefl a,
+  trans := to_lex.surjective.forall₃.2 $ λ a b c, @trans _ _ i.to_is_trans a b c }
+
+variables [linear_order α]
+
+/--  The partial order on `finsupp`s obtained by the lexicographic ordering.
+
+See `finsupp.lex.linear_order` for a proof that this partial order is in fact linear. -/
+instance lex.partial_order [partial_order N] : partial_order (lex (α →₀ N)) :=
+partial_order.lift (λ x, to_lex ⇑(of_lex x)) finsupp.coe_fn_injective--fun_like.coe_injective
+
+variable [linear_order N]
+
+/-  The "decidable fields" of `lex.linear_order` are proved by appealing to `classical` reasoning.
+It may be possible to get a constructive version of this instance, but this seems to require a
+possibly long detour and I (DT) am not sure of the details. -/
+/--  The linear order on `finsupp`s obtained by the lexicographic ordering. -/
+noncomputable instance lex.linear_order : linear_order (lex (α →₀ N)) :=
+{ le_total := to_lex.surjective.forall₂.2 $ λ f g, begin
+    cases (f.diff g).eq_empty_or_nonempty with he he,
+    { exact or.inl (finsupp.diff_eq_empty.mp he).le },
+    { cases he with a ha,
+      haveI : inhabited α := ⟨a⟩,
+      cases le_or_lt (of_lex f (f.wit g)) (of_lex g (f.wit g)) with mf mg,
+      { refine or.inl (or.inr ⟨f.wit g, λ j hj, apply_eq_of_le_wit hj, mf.lt_of_ne _⟩),
+        exact wit_eq_wit_iff.not.mpr (nonempty_diff_iff.mp ⟨_, ha⟩) },
+      { exact or.inr (or.inr ⟨g.wit f, λ j hj, apply_eq_of_le_wit hj, (by rwa wit_comm at mg)⟩) } }
+    end,
+  decidable_le := by { classical, apply_instance },
```

Comment:
```suggestion
  decidable_le := by { classical, apply_instance },
  decidable_eq := by apply_instance,
```
otherwise this creates an instance diamond